### PR TITLE
zynq7000-uart: launch uart instance as a separate process

### PR DESCRIPTION
**Pull request by HBuczynski**
## Description

UART number is taken from args. 

## Motivation and Context

Current implementation sets the same baudrate for all uarts.
In the new approach, each uart instance is in a separate process.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: tty/zynq7000-uart/zynq7000-uart.c [hunk 1]
```diff
@@ -455,45 +416,58 @@ int main(int argc, char **argv)
...
-		if (uart_init(n, baud) < 0)
-			continue;
+	if (uart_init(uartn, baud) < 0) {
+		debug("uart: cannot initialize uart\n");
+		return EXIT_FAILURE;
+	}
 
-		uart = &uart_common.uarts[n];
-		uart->oid.port = port;
+	uart = &uart_common.uart;
```
Comment (agkaminski) [inline tty/zynq7000-uart/zynq7000-uart.c:463]: This variable is not really needed

File: tty/zynq7000-uart/zynq7000-uart.c [hunk 2]
```diff
@@ -455,45 +416,58 @@ int main(int argc, char **argv)
...
 	}
 
-	uart_initClk();
+	if (uartn == (unsigned int)-1) {
+		debug("uart: device number is not provided\n");
+		return EXIT_FAILURE;
+	}
+
+	uart_initClk(uartn);
 	portCreate(&port);
```
Comment (agkaminski) [inline tty/zynq7000-uart/zynq7000-uart.c:456]: No retval checks

File: tty/zynq7000-uart/zynq7000-uart.c [hunk 3]
```diff
@@ -455,45 +416,58 @@ int main(int argc, char **argv)
...
-				break;
+	int c;
+	unsigned int uartn = -1;
+
+	if (argc == 1) {
+		uartn = 1;
+	}
+	else {
+		/* Get uart ID and baudrate */
+		while ((c = getopt(argc, argv, "n:b:")) != -1) {
```
Comment (agkaminski) [inline tty/zynq7000-uart/zynq7000-uart.c:427]: help/usage would be nice

File: tty/zynq7000-uart/zynq7000-uart.c [hunk 4]
```diff
@@ -455,45 +416,58 @@ int main(int argc, char **argv)
...
-				console = atoi(optarg);
-				if (console < 0 || console >= UARTS_MAX_CNT) {
-					debug("uart: wrong console ID\n");
-					return EXIT_FAILURE;
-				}
-				break;
-			default:
-				break;
+	int c;
+	unsigned int uartn = -1;
```
Comment (agkaminski) [inline tty/zynq7000-uart/zynq7000-uart.c:420]: You set it to 1 if argc == 1 anyway, no need for this

```suggestion
	unsigned int uartn = 1;
```

File: tty/zynq7000-uart/zynq7000-uart.c [hunk 5]
```diff
@@ -455,45 +416,58 @@ int main(int argc, char **argv)
...
+					break;
+			}
 		}
 	}
 
-	uart_initClk();
+	if (uartn == (unsigned int)-1) {
+		debug("uart: device number is not provided\n");
+		return EXIT_FAILURE;
+	}
```
Comment (agkaminski) [inline tty/zynq7000-uart/zynq7000-uart.c:453]:
```suggestion
```

File: tty/zynq7000-uart/zynq7000-uart.c [hunk 6]
```diff
@@ -455,45 +416,58 @@ int main(int argc, char **argv)
...
+				case 'b':
+					if ((baud = libtty_int_to_baudrate(atoi(optarg))) == (speed_t)-1) {
+						debug("uart: wrong baudrate value\n");
+						return EXIT_FAILURE;
+					}
+					break;
+
+				case 'n':
+					uartn = atoi(optarg);
+					if (uartn < 0 || uartn >= UARTS_MAX_CNT) {
```
Comment (agkaminski) [inline tty/zynq7000-uart/zynq7000-uart.c:438]: uartn cannot be < 0 - you declared it as unsigned.

File: tty/zynq7000-uart/zynq7000-uart.c [hunk 7]
```diff
@@ -375,8 +336,8 @@ static int uart_initClk(void)
 	/* Set IO PLL as source clock and set divider:
 	 * IO_PLL / 0x14 :  1000 MHz / 20 = 50 MHz     */
 	ctl.devclock.dev = pctl_ctrl_uart_clk;
-	ctl.devclock.clkact0 = 0x1;
-	ctl.devclock.clkact1 = 0x1;
+	ctl.devclock.clkact0 = (n == 0) ? 0x1 : 0;
+	ctl.devclock.clkact1 = (n == 1) ? 0x1 : 0;
```
[the remark this answers is not in the document]
Comment (HBuczynski) [reply tty/zynq7000-uart/zynq7000-uart.c]: Done

**Review by damianloew — COMMENTED** [review]
Verified that running `uart0` with 9600 `br` using  `X /sbin/zynq7000-uart -n 0 -b 9600` command placed in `rc.psh` works fine (communication with gps module, where supported `br` is 9600).

**Review by agkaminski — APPROVED** [review]

**Pull request merged**